ddtrace/span.py

```diff
@@ -168,6 +169,14 @@ def duration(self):
     def duration(self, value):
         self.duration_ns = value * 1e9
 
+    @property
+    def service(self):
+        return self._service
+
+    @service.setter
+    def service(self, name):
+        self._service = config.service_mapping.get(name, name)
```

Comment:
I wonder if this is the best place for this... this is definitely the most logical place for it.

What I am thinking though is, what is the impact of having getter/setter for service?

The service name will get set once per span, so that is a `dict().get()` getting called for every span we create. The service name will also get read for every child span created (It'll inherit it's parent's service name if it doesn't set one explicitly).

We could do this replacement at encode/flush time, but then we'd still have to incur a lookup cost on every single span, we are just moving it to a different process.


I don't think I am suggesting any kind of change here just yet, just some random musings, we should add a benchmark to measure the impact of this change. (We have some existing ones, just they might not cover a case that we care about for this change).

Reply:
I think regardless of where the support for the mapping is introduced, it is likely to have an impact on runtime performance. As you say, either we eagerly map the value, or take a more lazy approach; if all the spans are emitted anyway then the two approaches could have the same impact.

What was more of a concern for me is whether it's ok to store the mapped value on the `Span` object itself, or defer it until it's pushed out. With the current implementation, anybody accessing the `service` attribute will receive the mapped value back, which could not match the value that was passed to the constructor. The question is then whether anybody is relying on the assumption that `span.service` should match the value passed to the constructor or not. Given that the test suite was happy with my current implementation, I'm inclined to think that nobody is relying on such assumption, but I might be wrong.

Reply:
Another place we could do this, which might be more consistent with the implementation, is in `trace_utils.int_service` and `trace_utils.ext_service`. Only downside to that I see is that those utils aren't used by all of our integrations yet, but they will eventually.

Reply:
Would that be helpful for manual tracing?

Reply:
Ah true, I was thinking that the mapping would only apply to integrations, but it could just as well apply to custom instrumentation.

Perhaps then `tracer.start_span` is a worth considering since the remaining service naming logic there too.

Reply:
I like the idea of moving the mapping to `tracer.start_span`, if we all agree; this way any `Span` object won't give surprising results for the `service` attribute.

Reply:
+1 to moving to `tracer.start_span`, keeping the `Span` to more of a dumb container is good.